### PR TITLE
More robust fallback for `use` suggestion 

### LINKED ISSUE #87613: Strange `async use ...;` compiler suggestion
Really minor issue, but when building the following, I get a strange fix suggestion which seems incorrect:

```rust
#[tokio::main]
async fn main() {
    Command::new("git");
}
```

```toml
[package]
name = "replicate"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tokio = { version = "1", features = ["process", "rt-multi-thread", "macros"] }
```

```
   Compiling replicate v0.1.0 (/home/pleb/replicate)
error[E0433]: failed to resolve: use of undeclared type `Command`
 --> src/main.rs:3:5
  |
3 |     Command::new("git");
  |     ^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
2 | async use std::process::Command;
  |       ^^^^^^^^^^^^^^^^^^^^^^^^^^
2 | async use tokio::process::Command;
  |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0433`.
error: could not compile `replicate` due to previous error
```

```
pleb@gamey ~/replicate $ rustc --version
rustc 1.56.0-nightly (492723897 2021-07-29)
pleb@gamey ~/replicate $ rustup default
nightly-x86_64-unknown-linux-gnu (default)
```

The `async use` seems to be a bit strange for the suggestion. Nevertheless, have a wonderful day <3

### PR DESCRIPTION
**Pull request by pnkfelix**
Our old way to suggest where to add `use`s would first look for pre-existing `use`s in the relevant crate/module, and if there are *no* uses, it would fallback on trying to use another item as the basis for the suggestion.

But this was fragile, as illustrated in issue #87613 

This PR instead identifies span of the first token after any inner attributes, and uses *that* as the fallback for the `use` suggestion.

Fix #87613

### CODE REVIEW

**Comment by pnkfelix** [comment]
There is still one case that this arguably doesn't handle perfectly, namely this:

```rust
/* hey
whazzup */ fn main() {
    Command::new("git outta here");
}
```

where this *still* suggests:

```
2 | whazzup */ use std::process::Command;
  |            ++++++++++++++++++++++++++
```

(but that suggestion isn't *wrong*, and I expect that case to be absurdly rare, so I think its reasonable to just go ahead with this approach.)

**Comment by calebcartwright** [comment]
re: The rustfmt test failure, I'd think a fmt/tidy run would sort it.

That particular test is a bootstrapping sort for rustfmt which runs rustfmt against itself. Technically that means in this repository that test runs with a slightly different version of rustfmt than what x.py has pinned, so let me know if there's anything strange/conflicting with the formatting there.

**Comment by ehuss** [comment]
Nice!

There is a duplicate use placement finder at [`find_use_placement`](https://github.com/rust-lang/rust/blob/6d7684101a51f1c375ec84aef5d2fbdeb214bbc2/compiler/rustc_typeck/src/check/method/suggest.rs#L1939). Is it feasible to keep the implementations in sync?

Can you also update [`use_suggestion_placement.rs`](https://github.com/rust-lang/rust/blob/master/src/test/ui/resolve/use_suggestion_placement.rs#L28) to remove the FIXME?  Essentially something like:

```diff
--- a/src/test/ui/resolve/use_suggestion_placement.rs
+++ b/src/test/ui/resolve/use_suggestion_placement.rs
@@ -10,11 +10,7 @@ mod m {
 }

 mod foo {
-    // FIXME: UsePlacementFinder is broken because active attributes are
-    // removed, and thus the `derive` attribute here is not in the AST.
-    // An inert attribute should work, though.
-    // #[derive(Debug)]
-    #[allow(warnings)]
+    #[derive(Debug)]
     pub struct Foo;

     // test whether the use suggestion isn't
```

**Comment by pnkfelix** [comment]
> There is a duplicate use placement finder at [`find_use_placement`](https://github.com/rust-lang/rust/blob/6d7684101a51f1c375ec84aef5d2fbdeb214bbc2/compiler/rustc_typeck/src/check/method/suggest.rs#L1939). Is it feasible to keep the implementations in sync?

Oh, I don't know! I didn't know about that, but it seems obvious that I should try to do that.

> Can you also update [`use_suggestion_placement.rs`](https://github.com/rust-lang/rust/blob/master/src/test/ui/resolve/use_suggestion_placement.rs#L28) to remove the FIXME? Essentially something like:

Will do!

**Comment by pnkfelix** [comment]
> Nice!
> 
> There is a duplicate use placement finder at [`find_use_placement`](https://github.com/rust-lang/rust/blob/6d7684101a51f1c375ec84aef5d2fbdeb214bbc2/compiler/rustc_typeck/src/check/method/suggest.rs#L1939). Is it feasible to keep the implementations in sync?

Okay, I looked into doing this.

Unfortunately, the code I'm adding here is using the AST, but by the time typeck runs, the AST is long gone.

I don't want to block this PR on figuring out the best way to bring these two parts into sync. (E.g. one could thread down the same span information that I attached to the AST.) I will add a fixme at least, though.

**Comment by pnkfelix** [comment]
@calebcartwright can you explain this part of the [rustfmt formatting error](https://github.com/rust-lang/rust/pull/94584#issuecomment-1059587096) to me? (see below)

```diff
Mismatch at src/parse/parser.rs:113:
         let result = catch_unwind(AssertUnwindSafe(|| {
             let mut parser = new_parser_from_file(sess.inner(), path, Some(span));
             match parser.parse_mod(&TokenKind::Eof) {
-                Ok((a,
+                    a,

                     ast::ModSpans {
                         inner_span,
                         inner_span,

```

namely, why on earth is the diff saying that its *removing* `Ok((` from the text?

Was there a line that got removed in the report? I would have *perhaps* expected to see:

```
             match parser.parse_mod(&TokenKind::Eof) {
-                Ok((a,
+                Ok((
+                    a,
                     i,
```

**Comment by calebcartwright** [comment]
It does look a bit odd @pnkfelix, I haven't noticed it before. Worth noting though that these tests are using an internal, rudimentary diff emitter because of where/how the test is constructed, and it's not the same emitter and output for standard end user invocations of `rustfmt`.

May take a look at this if we get time, but not too concerned given it's only seen by the small audience of those modifying rustfmt source. Here's what an end user would see and what I would expect:

```
             match parser.parse_mod(&TokenKind::Eof) {
-                Ok((a, i, ast::ModSpans { inner_span, inject_use_span: _ })) => Some((a, i, inner_span)),
+                Ok((
+                    a,
+                    i,
+                    ast::ModSpans {
+                        inner_span,
+                        inject_use_span: _,
+                    },
+                )) => Some((a, i, inner_span)),

```

File: src/test/ui/proc-macro/amputate-span.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// aux-build:amputate-span.rs
```
Comment (estebank) [inline src/test/ui/proc-macro/amputate-span.rs:1]: Should we try and make this one `run-rustfix`?

**Review by estebank — APPROVED** [review]
r=me with or without the change to the test.

The follow up work to keep the suggestion span in the hir might be a good first-timer issue.

File: src/test/ui/resolve/use_suggestion_placement.fixed [hunk 2]
```diff
@@ -14,13 +14,9 @@ mod m {
 }
 
 mod foo {
-    // FIXME: UsePlacementFinder is broken because active attributes are
-    // removed, and thus the `derive` attribute here is not in the AST.
-    // An inert attribute should work, though.
-    // #[derive(Debug)]
```
Comment (camelid) [inline src/test/ui/resolve/use_suggestion_placement.fixed:20]: This bug has driven me nuts for a really long time, so thanks for fixing it ❤️

File: src/test/ui/proc-macro/amputate-span.rs (hunk 1 above)
Comment (pnkfelix) [reply src/test/ui/proc-macro/amputate-span.rs:1]: done in follow up commit, thank you for pointing that out!

**Pull request merged**